pkg/cmd/controller/controller_build.go

```diff
@@ -195,56 +194,40 @@ func (o *ControllerBuildOptions) Run() error {
 		return errors.Wrap(err, "failed to label the PipelineActivity resources")
 	}
 
+	var handlers watcher.HandlerFuncs
 	if tektonEnabled {
-		pod := &corev1.Pod{}
-		log.Logger().Infof("Watching for Pods in namespace %s", util.ColorInfo(ns))
-		listWatch := cache.NewListWatchFromClient(kubeClient.CoreV1().RESTClient(), "pods", ns, fields.Everything())
-		kube.SortListWatchByName(listWatch)
-		_, controller := cache.NewInformer(
-			listWatch,
-			pod,
-			time.Minute*10,
-			cache.ResourceEventHandlerFuncs{
-				AddFunc: func(obj interface{}) {
-					o.onPipelinePod(obj, kubeClient, jxClient, tektonClient, ns)
-				},
-				UpdateFunc: func(oldObj, newObj interface{}) {
-					o.onPipelinePod(newObj, kubeClient, jxClient, tektonClient, ns)
-				},
-				DeleteFunc: func(obj interface{}) {
-				},
+		log.Logger().Infof("Watching for tekton Pods in namespace %s", util.ColorInfo(ns))
+		handlers = watcher.HandlerFuncs{
+			AddFunc: func(obj runtime.Object) {
+				o.onPipelinePod(obj, kubeClient, jxClient, tektonClient, ns)
 			},
-		)
-
-		stop := make(chan struct{})
-		go controller.Run(stop)
+			ModifyFunc: func(obj runtime.Object) {
+				o.onPipelinePod(obj, kubeClient, jxClient, tektonClient, ns)
+			},
+			DeleteFunc: func(obj runtime.Object) {
+			},
+		}
 	} else {
-		pod := &corev1.Pod{}
 		log.Logger().Infof("Watching for Knative build pods in namespace %s", util.ColorInfo(ns))
-		listWatch := cache.NewListWatchFromClient(kubeClient.CoreV1().RESTClient(), "pods", ns, fields.Everything())
-		kube.SortListWatchByName(listWatch)
-		_, controller := cache.NewInformer(
-			listWatch,
-			pod,
-			time.Minute*10,
-			cache.ResourceEventHandlerFuncs{
-				AddFunc: func(obj interface{}) {
-					o.onPod(obj, kubeClient, jxClient, ns)
-				},
-				UpdateFunc: func(oldObj, newObj interface{}) {
-					o.onPod(newObj, kubeClient, jxClient, ns)
-				},
-				DeleteFunc: func(obj interface{}) {
-				},
+		handlers = watcher.HandlerFuncs{
+			AddFunc: func(obj runtime.Object) {
+				o.onPod(obj, kubeClient, jxClient, ns)
 			},
-		)
-
-		stop := make(chan struct{})
-		go controller.Run(stop)
+			ModifyFunc: func(obj runtime.Object) {
+				o.onPod(obj, kubeClient, jxClient, ns)
+			},
+			DeleteFunc: func(obj runtime.Object) {
+			},
+		}
 	}
 
-	// Wait forever
-	select {}
+	watcher := watcher.NewPodWatcher(kubeClient, ns, metav1.ListOptions{}, handlers)
+	for {
+		err = watcher.Run()
+		if err != nil {
```

Comment:
I think this loop retries endlessly. I might be useful to handle the case when an error is fatal and crash the pod to let the kubernetes scheduler to restart it.

Reply:
the tricky bit is knowing which errors are fatal and which are not - thats why I went with `log.Fatal` - to make explicit fatal errors

Reply:
the loop is designed loop endlessly - to recreate lists/watches and process the watches until watch errors - and terminate with `log.Fatal` if the watcher cannot be created

Reply:
I think you can define some custom error types in watcher package such as `FatalError` and `RetryableError` and wrap all error accordantly. On the caller side you can use a type switch to handle each case:
```
 switch e := err.(type) {
    case *RetryableError:
        // retry
    case *FatalError:
        // crash the pod
    default:
    }
```

Reply:
This will give the caller the chance to handle these cases in its own way. Maybe we want to reuse this code in other areas which want to handle this cases differently.